src/emqx_delayed_publish.erl

```diff
@@ -128,6 +142,20 @@ code_change(_OldVsn, State, _Extra) ->
 %% Internal functions
 %%------------------------------------------------------------------------------
 
+handle_message(ClientId, Delay, Msg = #message{id = Id, timestamp = Ts}) ->
```

Comment:
IMHO. a better abstraction is to move he Delay or Timestamp logic to a separate function (which can be easily eunit-tested).
But not to make a higer level abstraction as `handle_message`
i.e. the `ok = store(...),` call can be kept at its old place, so the return value  `{stop, Msg}` tuple do not have to be scattered into multiple functions.